design/service-binding.md

```diff
@@ -0,0 +1,851 @@
+
+## Binding data
+
+To connect to Strimzi, a service binding needs the followng:
+
+* **host** - hostname or IP address
+* **port** - port number
+* **userName** - username, optional
+* **password** - password or token, optional
+* **certificate** - certificate, optional
+
+These pieces of information can be provided in a variety of ways:
+
+* annotations on custom resources, indicating `spec` and `status` properties
+* ConfigMap
+* Secret
+
+## Binding to a Kafka cluster with no TLS or authentication
+
+Because Strimzi support multiple listeners and there is also a future plan to enhance the listener configuration capabilities, it seems prudent to design a scheme that works nicely with multiple listeners.
+
+There are two proposals here, with a preference for the first. Consequently, this format is used for the later examples.
+
+### Option 1 - Augment Kafka CR status
+
+This introduces a list of bootstrap addresses into the `status` for a `Kafka` CR like this:
+
+``` yaml
+apiVersion: kafka.strimzi.io/v1beta1
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+ ...
+status:
+  listeners:
+  - type: plain
+    addresses:
+    - host: myhost1.example.com
+      port: 9092
+    - host: myhost2.example.com
+      port: 9092
+  - type: tls
+    addresses:
+    - host: myhost3.example.com
+      port: 9093
+  bootstrap:
+    plain: myhost1.example.com:9092,myhost2.example.com:9092
+    tls: myhost3.example.com
+```
+
+This has the advantage that the bootstrap servers information is at a known point in the `status` of the `Kafka` custom resource which makes is simple to annotate the CSV so the Service Binding Operator can find it.
+
+Because that structure is a little odd, perhaps this would be preferred.
+
+``` yaml
+bootstrap:
+- name: plain
+  value: myhost1.example.com:9092,myhost2.example.com:9092
+- name: tls
+  value: myhost3.example.com
+```
+
+This is trying of course to find a syntax that is sufficiently general that the Service Binding Operator can be extended to support it.
+
+``` yaml
+apiVersion: operators.coreos/com:v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  name: strimzi-cluster-operator.v0.16.2
+  namespace: placeholder
+  annotations:
+    containerImage: docker.io/strimzi/operator:0.16.2
+spec:
+  customresourcedefinitions:
+    owned:
+    - description: Represents a Kafka cluster
+      displayName: Kafka
+      kind: Kafka
+      name: kafkas.kafka.strimzi.io
+      version: v1beta1
+      statusDescriptors:
+      - description: The addresses of the bootstrap servers for Kafka clients
+        displayName: Bootstrap servers
+        path: status.bootstrap
+        x-descriptors:
+        - binding:env:object:secret:endpoints
+```
+
+#### Consuming client's ServiceBindingRequest
+
+All of the required annotations are applied to the `Kafka` CSV, so the binding should only need to refer to the `Kafka` CR.
```

Comment:
_applied to the `Kafka` CSV_ ... CSV or CR? The annotations were on the Kafka CR in the past, the CSV has the UI descriptor updated. I'm not sure this is a typo or I missed something.

Reply:
The SBO looks at the CSV x-descriptors. They tell it how to extract the binding information from the CR status.